File: apps/mobile/modules/pdf-text-extractor/ios/PdfTextExtractorModule.swift
import ExpoModulesCore
import PDFKit
import os

public class PdfTextExtractorModule: Module {
  private let defaultMaxPages = 100

  public func definition() -> ModuleDefinition {
    Name("PdfTextExtractor")

    // 异步函数：提取 PDF 文本
    AsyncFunction("extractText") { (filePath: String, options: ExtractOptions?) -> [String: Any] in
      return try self.extractTextFromPDF(filePath: filePath, options: options)
    }

    // 异步函数：获取 PDF 页数
    AsyncFunction("getPageCount") { (filePath: String) -> Int in
      return self.getPageCount(filePath: filePath)
    }
  }

  // MARK: - PDF 文本提取核心逻辑

  private func extractTextFromPDF(filePath: String, options: ExtractOptions?) throws -> [String: Any] {
    let url = try parseFileURL(filePath)
    let didAccess = url.startAccessingSecurityScopedResource()
    defer {
      if didAccess {
        url.stopAccessingSecurityScopedResource()
      }
    }

    guard let document = PDFDocument(url: url) else {
      throw FailedToLoadDocumentException()
    }

    if document.isEncrypted {
      throw FailedToLoadDocumentException()
    }

    let totalPages = document.pageCount

    let maxPages = options?.maxPages ?? defaultMaxPages
    let endPage = min(maxPages, totalPages)
    let isTruncated = endPage < totalPages

    var extractedText = ""
    var extractionError = false

    for pageIndex in 0..<endPage {
      autoreleasepool {
        if let page = document.page(at: pageIndex) {
          if let pageText = page.string {
            if !extractedText.isEmpty {
              extractedText += "\n"
            }
            extractedText += pageText
          }
        } else {
          extractionError = true
        }
      }
    }

    return [
      "text": extractedText,
      "totalPages": totalPages,
      "extractedPages": endPage,
      "isTruncated": isTruncated,
      "extractionError": extractionError
    ]
  }

  private func getPageCount(filePath: String) -> Int {
    do {
      let url = try parseFileURL(filePath)
      let didAccess = url.startAccessingSecurityScopedResource()
      defer {
        if didAccess {
          url.stopAccessingSecurityScopedResource()
        }
      }
      guard let document = PDFDocument(url: url) else {
        os_log("PdfTextExtractor: failed to load PDF at %{public}@", log: .default, type: .error, filePath)
        return 0
      }
      return document.pageCount
    } catch {
      os_log("PdfTextExtractor: getPageCount error for %{public}@: %{public}@", log: .default, type: .error, filePath, String(describing: error))
      return 0
    }
  }

  // MARK: - 辅助方法

  private func parseFileURL(_ filePath: String) throws -> URL {
    let url: URL

    if filePath.hasPrefix("file://") {
      guard let parsed = URL(string: filePath) else {
        throw InvalidFilePathException()
      }
      url = parsed
    } else {
      url = URL(fileURLWithPath: filePath)
    }

    guard FileManager.default.fileExists(atPath: url.path) else {
      throw FileNotFoundException()
    }

    return url
  }

}

// MARK: - 数据结构

struct ExtractOptions: Record {
  @Field
  var maxPages: Int?
}

// MARK: - 错误定义


internal class InvalidFilePathException: Exception {
  override var code: String { "INVALID_FILE_PATH" }
  override var reason: String { "Invalid file path provided" }
}

internal class FileNotFoundException: Exception {
  override var code: String { "FILE_NOT_FOUND" }
  override var reason: String { "PDF file not found at the specified path" }
}

internal class FailedToLoadDocumentException: Exception {
  override var code: String { "FAILED_TO_LOAD_DOCUMENT" }
  override var reason: String { "Failed to load PDF document. The file may be corrupted or password-protected" }
}

File: apps/mobile/modules/pdf-text-extractor/android/src/main/java/expo/modules/pdftextextractor/PdfTextExtractorModule.kt
package expo.modules.pdftextextractor

import android.content.Context
import android.net.Uri
import com.tom_roush.pdfbox.android.PDFBoxResourceLoader
import com.tom_roush.pdfbox.pdmodel.PDDocument
import com.tom_roush.pdfbox.text.PDFTextStripper
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.io.File

class ExtractOptions : Record {
    @Field
    val maxPages: Int? = null
}

class InvalidFilePathException(cause: Throwable? = null) : CodedException(
    "INVALID_FILE_PATH",
    "Invalid file path provided",
    cause
)

class FileNotFoundException(cause: Throwable? = null) : CodedException(
    "FILE_NOT_FOUND",
    "PDF file not found at the specified path",
    cause
)

class FailedToLoadDocumentException(cause: Throwable? = null) : CodedException(
    "FAILED_TO_LOAD_DOCUMENT",
    "Failed to load PDF document. The file may be corrupted or password-protected",
    cause
)

class PdfTextExtractorModule : Module() {
    private val defaultMaxPages = 100
    @Volatile
    private var isInitialized = false

    private val context: Context
        get() = requireNotNull(appContext.reactContext)

    override fun definition() = ModuleDefinition {
        Name("PdfTextExtractor")

        OnCreate {
            initializePdfBox()
        }

        AsyncFunction("extractText") { filePath: String, options: ExtractOptions? ->
            extractTextFromPDF(filePath, options)
        }

        AsyncFunction("getPageCount") { filePath: String ->
            getPageCount(filePath)
        }
    }

    private fun initializePdfBox() {
        if (!isInitialized) {
            PDFBoxResourceLoader.init(context)
            isInitialized = true
        }
    }

    private fun extractTextFromPDF(filePath: String, options: ExtractOptions?): Map<String, Any> {
        val file = parseFilePath(filePath)
        val isTempFile = filePath.startsWith("content://")

        try {
            val document: PDDocument = try {
                PDDocument.load(file)
            } catch (e: Exception) {
                throw FailedToLoadDocumentException(e)
            }

            return document.use { doc ->
                val totalPages = doc.numberOfPages

                val maxPages = options?.maxPages ?: defaultMaxPages
                val endPage = minOf(maxPages, totalPages)
                val isTruncated = endPage < totalPages
                if (totalPages == 0) {
                    return mapOf(
                        "text" to "",
                        "totalPages" to 0,
                        "extractedPages" to 0,
                        "isTruncated" to false,
                        "extractionError" to false
                    )
                }

                val stripper = PDFTextStripper().apply {
                    startPage = 1
                    this.endPage = endPage
                }

                var extractionError = false
                val extractedText = try {
                    stripper.getText(doc)
                } catch (e: Exception) {
                    android.util.Log.w("PdfTextExtractor", "Failed to extract text from PDF", e)
                    extractionError = true
                    ""
                }

                mapOf(
                    "text" to extractedText,
                    "totalPages" to totalPages,
                    "extractedPages" to endPage,
                    "isTruncated" to isTruncated,
                    "extractionError" to extractionError
                )
            }
        } finally {
            if (isTempFile && file.exists()) {
                file.delete()
            }
        }
    }

    private fun getPageCount(filePath: String): Int {
        return try {
            val isTempFile = filePath.startsWith("content://")
            val file = parseFilePath(filePath)
            try {
                PDDocument.load(file).use { doc ->
                    doc.numberOfPages
                }
            } finally {
                if (isTempFile && file.exists()) {
                    file.delete()
                }
            }
        } catch (e: Exception) {
            0
        }
    }

    private fun parseFilePath(filePath: String): File {
        val file = when {
            filePath.startsWith("file://") -> {
                val uri = Uri.parse(filePath)
                File(uri.path ?: throw InvalidFilePathException())
            }
            filePath.startsWith("content://") -> {
                val uri = Uri.parse(filePath)
                copyContentUriToTempFile(uri)
            }
            else -> File(filePath)
        }

        if (!file.exists()) {
            throw FileNotFoundException()
        }

        return file
    }

    private fun copyContentUriToTempFile(uri: Uri): File {
        val tempFile = File.createTempFile("pdf_temp_", ".pdf", context.cacheDir)

        try {
            val inputStream = context.contentResolver.openInputStream(uri)
                ?: throw InvalidFilePathException()
            inputStream.use { input ->
                tempFile.outputStream().use { output ->
                    input.copyTo(output)
                }
            }
        } catch (e: Exception) {
            if (tempFile.exists()) {
                tempFile.delete()
            }
            throw e
        }

        return tempFile
    }
}
